Add rudimentry debug info metadata emission

dwblaikie opened the pull request:
This adds just the debug info metadata for Compilation Units (the top level container of debug info) - but without anything in them, LLVM won't emit them at all, so while this is testable at the IR level, it isn't observable at the object level until more debug info is added.

A couple of starting points in this patch:
* A flag (`--debug-info`, seems to match the naming/style of other flags in the carbon driver, though this is different from the naming conventions of clang/gcc) that enables debug info when lowering. Open to other names/approaches (on by default?  historically debug info's been to large/expensive to do this, so sticking with that precedent for now).
* Enabling that flag by default in the lowering tests - I do find the churn on golden tests a bit rough, and adding more features to all the tests means more churn, but it seems consistent with the approach so far - keep an eye on this and perhaps revisit this if the churn gets too annoying

dwblaikie committed:
Add rudimentry debug info metadata emission

This adds just the debug info metadata for Compilation Units (the top
level container of debug info) - but without anything in them, LLVM
won't emit them at all, so while this is testable at the IR level, it
isn't observable at the object level until more debug info is added.

jonmeow commented on toolchain/lower/file_context.h:
```diff
@@ -101,6 +107,9 @@ class FileContext {
   llvm::SmallVector<llvm::Constant*, 0> constants_;
+
+  // Specify whether to include debug info metadata in the generated LLVM IR.
+  bool include_debug_info_;
```
WDYT of moving this up in the member list, maybe before `sem_ir` so that the member order and constructor parameter order correspond a little more?

dwblaikie replied:
Sure - though ended up removing it once `BuildDICompileUnit` was a standalone helper, and we can rely on the CU being non-null as the signifier of whether debug info is being emitted when later patches add other debug info emission further along the lowering pipeline.

jonmeow commented on toolchain/lower/file_context.cpp:
```diff
@@ -66,6 +70,22 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
+  if (!include_debug_info_)
+    return;
+  // FIXME: Include directory path in the cu_file.
+  llvm::DIFile* cu_file = di_builder_.createFile(module_name, "");
```
Pardon my unfamiliarity, but why "cu" instead of "di"?

dwblaikie replied:
Ah - because it's the file for the compilation unit (cu) - I'll expand out that here. These domain-specific acronyms make more sense in other parts of LLVM/Clang that are more focussed just on debug info, but here where it's mixed with non-debug-info concepts, being more explicit is probably good.

jonmeow commented on toolchain/lower/file_context.cpp:
```diff
@@ -20,16 +20,20 @@
+      include_debug_info_(include_debug_info) {
   CARBON_CHECK(!sem_ir.has_errors())
       << "Generating LLVM IR from invalid SemIR::File is unsupported.";
+  BuildCompileUnit(module_name);
```
Maybe rename this to `BuildDICompileUnit`? I was having trouble seeing where `di_compile_unit_` was initialized -- but maybe it'd be even better to have the method be static and return the `DICompileUnit*` instead of void, allowing initializing as `di_compile_unit_(BuildDICompileUnit(module_name))`?

dwblaikie replied:
Sure, renamed to BuildDICompileUnit.
It'd take a few more parameters to make it static (`include_debug_info_` (though we could move that out into the call site), `di_builder_`, and `llvm_module_`)) - I'll give it a go and you can see whether it feels better that way.

jonmeow replied:
This looks good, I think one nice thing about static is it's harder to accidentally use uninitialized members.

jonmeow commented on toolchain/lower/file_context.cpp:
```diff
@@ -66,6 +70,22 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
+  // Probably have to bump to C++ at some point for virtual functions,
+  // templates, etc.
+  di_compile_unit_ = di_builder_.createCompileUnit(
+      llvm::dwarf::DW_LANG_C, cu_file, "carbon", false, "", 0, "");
```
Is there value in labelling the last 4 parameters? I think it may make it easier for someone like me to understand the parameters without going to docs. That said, it looks like the last defaults to empty, so maybe can be removed.

```suggestion
      llvm::dwarf::DW_LANG_C, cu_file, "carbon", /*isOptimized=*/false, /*Flags=*/"", /*RV=*/0);
```

jonmeow reviewed (commented):
Looks good! Thank you for the contributions. Just a few style comments.

Are you using [pre-commit](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md#running-pre-commit) (`pre-commit install`)? I think it would've automatically fixed this by running `clang-format` for you.

dwblaikie committed:
left-bind '*' in toolchain/lower/file_context.h

Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>

dwblaikie committed:
Add missing braces

Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>

dwblaikie committed:
add argument name comments for clarity

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

dwblaikie committed:
Rename BuildCompileUnit to BuildDICompileUnit

dwblaikie committed:
Make FileContext::BuildDICompileUnit a factory function

(also rename it, adding the 'DI' in the name to match the type and
member variable name)

This renders `include_debug_info_` unused, so remove it - we can rely on
the CU being non-null as the marker for debug info emission in future
patches/work.

dwblaikie committed:
Rename cu_file to compile_unit_file for clarity

dwblaikie committed:
clang-format

dwblaikie commented:
> Are you using [pre-commit](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md#running-pre-commit) (`pre-commit install`)? I think it would've automatically fixed this by running `clang-format` for you.

Ah, nope - missed that, sorry. Thanks for the pointer/reminder. Think I managed to run it on some of my stuff that's already committed (with `pre-commit run --all-files`) to the branch - it caught one formatting error, but otherwise seems clean

jonmeow reviewed (approved):
Thanks, glad to have debug info being added!

chandlerc commented:
> Mostly this'll be a place to discuss the flag naming/functionality, and testing strategy. Should we have debug info on in all lowering tests? Should there be some separate directory just for debug info lowering test coverage? I'm open to ideas/suggestions/requests.

Not a comment on the patch, but maybe update the description before merging to summarize what your suggested starting point is for these questions, and what motivates that starting point?

(FWIW, not disagreeing with the current behavior, mostly looking to have a record of it.)

dwblaikie committed:
Merge branch 'trunk' into debug_info_flag_and_compilation_unit

dwblaikie committed:
Update tests

dwblaikie commented:
> > Mostly this'll be a place to discuss the flag naming/functionality, and testing strategy. Should we have debug info on in all lowering tests? Should there be some separate directory just for debug info lowering test coverage? I'm open to ideas/suggestions/requests.
> 
> Not a comment on the patch, but maybe update the description before merging to summarize what your suggested starting point is for these questions, and what motivates that starting point?
> 
> (FWIW, not disagreeing with the current behavior, mostly looking to have a record of it.)

Updated - hopefully that's something along the lines of what you're looking for.

dwblaikie committed:
Add test for driver flag compatibility error

dwblaikie committed:
Autoupdate tests
